cpp/src/copying/contiguous_split.cu

```diff
@@ -14,466 +14,1129 @@
+ * When doing a contiguous_split on X columns comprised of N total internal buffers
+ * with M splits, we end up having to copy N*M source/destination buffer pairs.
+ * This kernel is arrangedsuch that each block copies 1 source/destination pair.
+ * This function retrieves the relevant function and then calls copy_buffer to perform
+ * the actual copy.
+ *
+ * @param num_src_bufs Total number of source buffers (N)
+ * @param num_partitions Number of partitions the each source buffer is split into (M)
+ * @param src_bufs Input source buffers (N)
+ * @param dst_bufs Desination buffers (N*M)
+ * @param buf_info Information on the range of values to be copied for each destination buffer.
+ *
+ * @returns Total offset stack size needed for this range of columns.
  */
-struct column_split_info {
-  size_t data_buf_size;      // size of the data (including padding)
-  size_t validity_buf_size;  // validity vector size (including padding)
+__global__ void copy_partition(int num_src_bufs,
+                               int num_partitions,
+                               uint8_t** src_bufs,
+                               uint8_t** dst_bufs,
+                               dst_buf_info* buf_info)
+{
+  int partition_index = blockIdx.x / num_src_bufs;
+  int src_buf_index   = blockIdx.x % num_src_bufs;
+  int t               = threadIdx.x;
+  size_t buf_index    = (partition_index * num_src_bufs) + src_buf_index;
+  int num_elements    = buf_info[buf_index].num_elements;
+  int element_size    = buf_info[buf_index].element_size;
+  int stride          = blockDim.x;
+  int src_row_index   = buf_info[buf_index].src_row_index;
+  uint8_t* src        = src_bufs[src_buf_index];
+  uint8_t* dst        = dst_bufs[partition_index] + buf_info[buf_index].dst_offset;
 
-  size_t offsets_buf_size;  // (strings only) size of offset column (including padding)
-  size_type num_chars;      // (strings only) number of chars in the column
-  size_type chars_offset;   // (strings only) offset from head of chars data
-};
+  // copy, shifting offsets and validity bits as needed
+  copy_buffer(dst,
+              src,
+              t,
+              num_elements,
+              element_size,
+              src_row_index,
+              stride,
+              buf_info[buf_index].value_shift,
+              buf_info[buf_index].bit_shift);
+}
+
+// The block of function below are all related:
+//
+// compute_offset_stack_size()
+// setup_src_buf_data()
+// count_src_bufs()
+// setup_source_buf_info()
+// build_output_columns()
+//
+// Critically, they all traverse the hierarchy of source columns and their children
+// in a specific order to guarantee they produce various outputs in a consistent
+// way.  For example, setup_src_buf_info() produces a series of information
+// structs that must appear in the same order that setup_src_buf_data() produces
+// buffers.
+//
+// So please be careful if you change the way in which these functions and
+// functors traverse the hierarchy.
```

Comment:
Comments like this make me wary. They give off a code smell that there are expectations being enforced only via convention instead of somehow being enforced via the code. E.g., perhaps there should be an algorithm that abstracts how the hierarchy is traversed and you can pass a lambda that customizes what you do during the traversal. That way you can ensure the traversal is always done in the same way for all functions.